src/super_gradients/training/utils/callbacks/callbacks.py

```diff
@@ -948,3 +953,205 @@ def create_lr_scheduler_callback(
         raise ValueError(f"Unknown lr_mode: {lr_mode}")
 
     return sg_lr_callback
+
+
+class ExtremeBatchCaseVisualizationCallback(Callback):
+    """
+    ExtremeBatchCaseVisualizationCallback
+
+    A base class for visualizing worst/best validation batches in an epoch
+     according to some metric or loss value, with Full DDP support.
+
+    Images are saved with training_hyperparams.sg_logger.
+
+    :param metric_name: str,will be the metric which the model checkpoint will be saved according to, and can be set to any
+      of the following:
+
+        a metric name (str) of one of the metric objects from the training_hyperparams.valid_metrics_list
+
+        a "component_name" if some metric in valid_metrics_list has an attribute component_names. In such cas it
+        is a list referring to the names of each entry in the output metric (torch tensor of size n).
+
+        one of "loss_logging_items_names" i.e which will correspond to an item returned during the
+        loss function's forward pass (see loss docs in Trainer.train(..)).
+
+    :param max: bool, Whether to take the batch corresponding to the max value of the metric/loss or
+     the minimum (default=False).
+
+    :param freq: int, epoch frequency to perform all of the above (default=1).
+
+     Inheritors should implement process_extreme_batch which returns an image, as an np.array (uint8) with shape BCHW.
+    """
+
+    def __init__(self, metric_name: str, max: bool = False, freq: int = 1):
+        self.metric_name = metric_name
+        self.metric = None
+        self.max = max
+        self.freq = freq
+        self.extreme_score = -1 * np.inf if max else np.inf
+
+        self.extreme_batch = None
+        self.extreme_preds = None
+        self.extreme_targets = None
+
+        self._first_call = True
+        self._idx_loss_tuple = None
+        self._tag = f"max_{self.metric_name}_batch" if self.max else f"min_{self.metric_name}_batch"
+
+        super(ExtremeBatchCaseVisualizationCallback, self).__init__()
+
+    def process_extreme_batch(self) -> np.array:
+        raise NotImplementedError
```

Comment:
I would rather have `@abstractmethod` (along with `ExtremeBatchCaseVisualizationCallback(Callback, ABC)` ) because this will raise an exception on instantiation and not just on call.
It will also help people implement new methods because it is more explicit and IDE detects it

Reply:
Agree, changed in 9bb4c9b0957b9cffb27e10a96450a0c47c5235ec